validator/network_enable_flow_logs.rego

```diff
@@ -30,9 +30,12 @@ deny[{
 	legacy_enable_flow_logs := lib.get_default(network, "enableFlowLogs", false)
 	log_config := lib.get_default(network, "logConfig", {})
 	log_config_enable_flow_logs := lib.get_default(log_config, "enable", false)
+	purpose := lib.get_default(network, "purpose", "PRIVATE")
 
 	log_config_enable_flow_logs != true
 	legacy_enable_flow_logs != true
+	purpose != "REGIONAL_MANAGED_PROXY"
+	purpose != "INTERNAL_HTTPS_LOAD_BALANCE"
```

Comment:
Shouldn't this be an "==" comparison since the flow logs will be disable IF the purpose IS set to `REGIONAL_MANAGED_PROXY` or `INTERNAL_HTTPS_LOAD_BALANCER` 

Also, I think this should be ORed as:
`any([purpose == "REGIONAL_MANAGED_PROXY", purpose == "INTERNAL_HTTPS_LOAD_BALANCER"])`

Reply:
Actually this may be bit more nuanced because if we determine that the flow logs are already set to false, we don't need to check for the purpose anymore. We would only check the purpose IF the flow logs were set to true.

Reply:
Hi @g-awmalik thanks for the review.
My thoughts on the points you highlighted

1) The semantics of the validation  check is to trigger a violation if **the following condition is `true`**: 

```
log_config_enable_flow_logs != true
legacy_enable_flow_logs != true
```

which could be rewritten as: 

- `log_config_enable_flow_logs` is not set to `true` **AND** `legacy_enable_flow_logs` is not set to `true` => rise **Violation**

with the change we will have:

```
log_config_enable_flow_logs != true
legacy_enable_flow_logs != true
purpose != "REGIONAL_MANAGED_PROXY"
purpose != "INTERNAL_HTTPS_LOAD_BALANCE"
```

which could be rewritten as: 

- `log_config_enable_flow_logs` is not set to `true` **AND** `legacy_enable_flow_logs` is not set to `true`
**AND** `purpose` is not `REGIONAL_MANAGED_PROXY` **AND** `purpose` is not `INTERNAL_HTTPS_LOAD_BALANCE` => rise **Violation**


2) on the other topic, of  a more nuanced validation, considering the case of subnetwork having `REGIONAL_MANAGED_PROXY` as the purpose we would have two cases:

- vpc flow log is NOT enabled and purpose is `REGIONAL_MANAGED_PROXY` => **No violation** is raised, and the resource configuration is OK
- vpc flow log is enabled and purpose is `REGIONAL_MANAGED_PROXY` => No violation is raised BUT the resource configuration is NOT OK (because if the purpose is `REGIONAL_MANAGED_PROXY` it should not have a vpc flow logs configuration) BUT this is one of the validations of the API and will rise an error in apply. this is not security recommendation issue.

Reply:
In that case we're not completely shifting left the validation. I.e. for a configuration that has `purpose` set as `REGIONAL_MANAGED_PROXY` and has flow logs enabled, there will be no violation which raises question if adding a purpose check is adding any value?

Reply:
do you mean that we should use

```
purpose != "REGIONAL_MANAGED_PROXY"
purpose != "INTERNAL_HTTPS_LOAD_BALANCE"
log_config_enable_flow_logs != true
legacy_enable_flow_logs != true
```

Reply:
But this doesn't catch the scenario where purpose is `REGIONAL_MANAGED_PROXY` or `INTERNAL_HTTPS_LOAD_BALANCER` and has flow logs enabled. That won't raise a violation but will return and error at the API level.

I don't know if that's even possible with rego and that's why I'm suggesting we don't make this change since it doesn't solve the problem completely. I'll have to consume a few brain cells to see if this is even possible in rego.